app/scripts/background.js

```diff
@@ -444,33 +476,35 @@ browser.runtime.onConnect.addListener(async (...args) => {
   } catch (error) {
     sentry?.captureException(error);
 
-    // if have a STATE_CORRUPTION_ERROR tell the user about it and offer to
+    // if we have a STATE_CORRUPTION_ERROR tell the user about it and offer to
     // restore from a backup, if we have one.
-    if (KNOWN_STATE_CORRUPTION_ERRORS.has(error.message)) {
-      const _state = await localStore.get().catch((_) => null);
-
-      const port = args[0];
-      try {
-        // send the `error` TO THE ui
-        port.postMessage({
-          data: {
-            method: METHOD_DISPLAY_STATE_CORRUPTION_ERROR,
-            params: {
-              error: {
-                message: error.message,
-                name: error.name,
-                stack: error.stack,
-              },
-              currentLocale: _state?.data?.PreferencesController?.currentLocale,
-            },
-          },
-        });
-      } catch (e) {
-        // an exception can occur here if the Window has since disconnected from
-        // the background, this is expected if the UI is closed while we are
-        // still initializing the background (during `await isInitialized;`)
-        log.error('MetaMask - Failed to send state corruption error', e);
-      }
+    if (isStateCorruptionError(error)) {
+      await corruptionHandler.handleStateCorruptionError(
+        args[0],
+        error,
+        persistenceManager,
+        async (backup) => {
+          // we are going to reinitialize the background script, so we need to
+          // reset the initialization promises. this is gross since it is
+          // possible the original references could have been passed to other
+          // functions, and we can't update those references from here.
+          // right now, that isn't the case though.
+          setGlobalInitializers();
+
+          if (hasVault(backup)) {
+            await initBackground(backup);
+            controller.onboardingController.setFirstTimeFlowType(
+              FirstTimeFlowType.restore,
+            );
+          } else {
+            // if we don't have a backup we need to make sure we clear the state
+            // from the database, and then reinitialize the background script
+            // with the first time state.
+            await persistenceManager.reset();
+            await initBackground();
+          }
+        },
+      );
```

Comment:
Previously, this just sent a message to the UI, it still does, but it now does it in a function in another file - it will handle much more complex logic and flows.

The function, `handleStateCorruptionError`, just handles the user communications. The reset part happens in the callback.

I did it this way because of all the global variables the functions in this callback interact with; it was difficult finding a point of clean separation between _handling_ state corruption and logic related to resetting and restarting MM.